contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+
+2. Unexpected umount behavior in kubelet. Kubelet will umount the device from node if no pods use a volume, the `UnmountDeviceFunc` will use the device path stored in `Node.Status.VolumesAttached` to check if this device is still opened or not. As this device path is empty for local volumes, there may be unexpected behaviors.
+
+3. Complex and confusing code both in kubernetes side(`kubelet Volumemanager` and `A/D controller`) and volume plugin side, as local volumes have to implement `AttachableVolumePlugin` interface.
+
+## <a name="object"></a>Object
+
+To solve the problems as above, we need to split the `AttachableVolumePlugin` interface into two independent interfaces: One is still called `AttachableVolumePlugin` and another is `DeviceMountableVolumePlugin` . The new `AttachableVolumePlugin` has the same interface with the old one and implemented by the cloud volumes only, `DeviceMountableVolumePlugin` has the device mount/umount interface only and implemented by the local volumes. Accordingly, we also need to split `Attacher` intreface to `Attacher` and `DeviceMounter`, `Detacher` to `Detacher` and `DeviceUmounter`.
+
+## <a name="roadmap"></a>Roadmap
+
+* Volume relative interface refactor
+	* Refactor volume plugin interfaces(`AttachableVolumePlugin`) and volume interfaces(`Attacher/Detacher`)
+	* Refactor `A/D controller` and `Kubelet VolumeManager` to use new interfaces
+* Refactor local volumes to implement `DevicemountableVolumePlugin` only
+	* Refactor `rbd`
+	* Refactor `fc`, `iscsi`, etc.
+
+## <a name="design"></a>Implementation Design
+
+### <a name="k8s_refactor"></a>Kubernetes Side Refactor
+
+#### <a name="k8s_interfaces_refactor"></a>Volume Plugin Interfaces refactor
+
+We extract the device mount relative methods from `Attacher` to a new interface, and embed this new interface back to `Attacher` so that volumes implement `Attacher` need also implement `DeviceMounter`:
+
+```go
+// Attacher can attach a volume to a node.
+type Attacher interface {
+	DeviceMounter
+
+	// Attaches the volume specified by the given spec to the node with the given Name.
+	// On success, returns the device path where the device was attached on the
+	// node.
+	Attach(spec *Spec, nodeName types.NodeName) (string, error)
+
+	// VolumesAreAttached checks whether the list of volumes still attached to the specified
+	// node. It returns a map which maps from the volume spec to the checking result.
+	// If an error is occurred during checking, the error will be returned
+	VolumesAreAttached(specs []*Spec, nodeName types.NodeName) (map[*Spec]bool, error)
+
+	// WaitForAttach blocks until the device is attached to this
+	// node. If it successfully attaches, the path to the device
+	// is returned. Otherwise, if the device does not attach after
+	// the given timeout period, an error will be returned.
+	WaitForAttach(spec *Spec, devicePath string, pod *v1.Pod, timeout time.Duration) (string, error)
+}
+
+// DeviceMounter can mount a volume to node.
```

Comment:
I'd emphasize that it's for volumes backed by a block device. It's useless for Gluster or NFS.